capsules/src/net/icmpv6/icmpv6.rs

```diff
@@ -0,0 +1,182 @@
+//! ICMPv6 layer of the Tock networking stack.
+//!
+//! - Author: Conor McAvity <cmcavity@stanford.edu>
+
+use net::stream::{decode_u16, decode_u32, decode_u8};
+use net::stream::{encode_u16, encode_u32, encode_u8};
+use net::stream::SResult;
+
+#[derive(Copy, Clone)]
+pub struct ICMP6Header {
+    pub code: u8,
+    pub cksum: u16,
+    pub options: ICMP6HeaderOptions,
+    pub len: u16, // Not ICMP field, here for convenience
+}
+
+#[derive(Copy, Clone)]
+pub enum ICMP6HeaderOptions {
+    Type1 { unused: u32 },
+    Type3 { unused: u32 },
+    Type128 { id: u16, seqno: u16 },
+    Type129 { id: u16, seqno: u16 },
+}
+
+#[derive(Copy, Clone)]
+pub enum ICMP6Type {
+    Type1,   // Destination Unreachable
+    Type3,   // Time Exceeded
+    Type128, // Echo Request
+    Type129, // Echo Reply
+}
+
+impl ICMP6Header {
+    pub fn new(icmp_type: ICMP6Type) -> ICMP6Header {
+        let options = match icmp_type {
+            ICMP6Type::Type1 => ICMP6HeaderOptions::Type1 { unused: 0 },
+            ICMP6Type::Type3 => ICMP6HeaderOptions::Type3 { unused: 0 },
+            ICMP6Type::Type128 => ICMP6HeaderOptions::Type128 { id: 0, seqno: 0 },
+            ICMP6Type::Type129 => ICMP6HeaderOptions::Type129 { id: 0, seqno: 0 },
+        };
+
+        ICMP6Header {
+            code: 0,
+            cksum: 0,
+            options: options,
+            len: 0,
+        }
+    }
+
+    pub fn set_type(&mut self, icmp_type: ICMP6Type) {
+        match icmp_type {
+            ICMP6Type::Type1 => self.set_options(ICMP6HeaderOptions::Type1 { unused: 0 }),
+            ICMP6Type::Type3 => self.set_options(ICMP6HeaderOptions::Type3 { unused: 0 }),
+            ICMP6Type::Type128 => self.set_options(ICMP6HeaderOptions::Type128 { id: 0, seqno: 0 }),
+            ICMP6Type::Type129 => self.set_options(ICMP6HeaderOptions::Type129 { id: 0, seqno: 0 }),
+        }
+    }
+
+    pub fn set_code(&mut self, code: u8) {
```

Comment:
What's the point of having these methods if the fields are `pub` and can be modified directly?

Reply:
(ditto for `cksum`, `options` and `len`)

Reply:
Fields should not be pub.

Reply:
Why?